### PR TITLE
Unknown edns

### PR DESCRIPTION
**Pull request by cmouse**
When receiving EDNS query with unknown version, make sure we:
- send BADVERS
- set OPT record version to 0
- do not send SOA

### CODE REVIEW

**Comment by cmouse** [comment]
Ready for review.

**Comment by cmouse** [comment]
@Habbie can you review this?

**Comment by ahupowerdns** [comment]
I'm a bit confused. There is rcode and extended rcode. This patch appears to not distinguish between these, and sets an rcode to '16', which I don't know what it is (BADVERS I expect). But are those only the upper bits of the rcode? The first 4 bits of the rcode go in the dnsheader, the rest in edns extension, right?

**Comment by cmouse** [comment]
http://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-6

I did not know this. Interesting way to encode RCodes.

**Comment by cmouse** [comment]
Fixed.

dig said ;; ->>HEADER<<- opcode: QUERY, status: BADVERS, id: 37823
so it seems to be compatible now.

**Comment by Habbie** [comment]
shouldn't replyPacket also deal with the two new variables?

**Comment by cmouse** [comment]
Can you please clarify?

**Comment by Habbie** [comment]
I see the two new variables being initialized in the constructor and in the copy constructor, but not in the replypacket constructor. I am asking whether that is okay.

**Comment by cmouse** [comment]
Ah. Must've missed that, good catch.

**Comment by cmouse** [comment]
Actually, WHY would we copy response code from Question to Reply? It gets initialized to no error in ctor anyways.

**Comment by Habbie** [comment]
Indeed it does!

File: regression-tests.nobackend/edns1/command [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+                echo "Domain parsing failed" >> failed_tests
+        fi
+}
+
+$RUNWRAPPER ../pdns/pdns_server --daemon=no --local-port=$port --socket-dir=./          \
+        --no-shuffle --launch=bind --bind-config=edns-packet-cache/named.conf   \
+        --send-root-referral --cache-ttl=60 --no-config --module-dir=../regression-tests/modules &
+bindwait
+
+./edns1/test-edns.py
```
Comment (Habbie) [inline regression-tests.nobackend/edns1/command]: the 3.4.6 version has a timeout and a comment here

File: pdns/dnspacket.cc [hunk 2]
```diff
@@ -101,6 +103,8 @@ DNSPacket::DNSPacket(const DNSPacket &orig)
   d_eso = orig.d_eso;
   d_haveednssubnet = orig.d_haveednssubnet;
   d_haveednssection = orig.d_haveednssection;
+  d_ednsversion = orig.d_ednsversion;
+  d_ednsrcode = 0;
```
Comment (Habbie) [inline pdns/dnspacket.cc]: why zero instead of orig?

Comment (cmouse) [inline pdns/dnspacket.cc]: fixed.

Comment (Habbie) [inline pdns/dnspacket.cc]: In that case replypacket may want to zero it

Comment (cmouse) [inline pdns/dnspacket.cc]: if we actually used copy ctor for it, then yes. we don't do it, but I'll add it just to be explicit about it.

Comment (Habbie) [inline pdns/dnspacket.cc]: oh, you're right. Good call indeed

**Pull request merged**